openquake/commonlib/calculators/event_based.py

```diff
@@ -0,0 +1,307 @@
+            (sources, self.sitecol, monitor),
+            concurrent_tasks=self.oqparam.concurrent_tasks,
+            weight=operator.attrgetter('weight'),
+            key=operator.attrgetter('trt_model_id'))
+
+    def post_execute(self, result):
+        for trt_id in sorted(result):
+            for sr in result[trt_id]:
+                for tag in sorted(sr.tags):
+                    print tag
+        return {}
+
+
+# ######################## GMF calculator ############################ #
+
+def compute_gmfs_and_curves(ses_ruptures, sitecol, gsims_assoc, monitor):
+    """
+    :param ses_ruptures:
+        a list of blocks of SESRuptures with homogeneous TrtModel
+    :param sitecol:
+        a :class:`openquake.hazardlib.site.SiteCollection` instance
+    :param gsims_assoc:
+        associations trt_model_id -> gsims
+    :param monitor:
+        a Monitor instance
+    :returns:
+        a dictionary trt_model_id -> curves_by_gsim
+        where the list of bounding boxes is empty
+    """
+    oq = monitor.oqparam
+
+    # NB: by construction each block is a non-empty list with
+    # ruptures of the trt_model_id
+    trt_id = ses_ruptures[0].trt_model_id
+    gsims = sorted(gsims_assoc[trt_id])
+    imts = map(from_string, oq.imtls)
+    trunc_level = getattr(oq, 'truncation_level', None)
+    correl_model = readinput.get_correl_model(oq)
+
+    result = AccumDict({(trt_id, gsim.__class__.__name__): ([], AccumDict())
+                        for gsim in gsims})
+    for rupture, group in itertools.groupby(
+            ses_ruptures, operator.attrgetter('rupture')):
+        srs = list(group)
+        indices = srs[0].site_indices
+        r_sites = (sitecol if indices is None else
+                   site.FilteredSiteCollection(indices, sitecol))
+        for gsim in gsims:
+            gsim_name = gsim.__class__.__name__
+            computer = calc.gmf.GmfComputer(
+                rupture, r_sites, imts, gsim, trunc_level, correl_model)
+            for sr in srs:
+                gmf_by_imt = AccumDict(computer.compute(sr.seed))
+                gmf_by_imt.tag = sr.tag
+                gmf_by_imt.r_sites = r_sites
+                result[trt_id, gsim_name][0].append(gmf_by_imt)
+
+    if getattr(oq, 'hazard_curves_from_gmfs', None):
+        duration = oq.investigation_time * oq.ses_per_logic_tree_path
+        for gsim in gsims:
+            gmfs, curves = result[trt_id, gsim.__class__.__name__]
+            curves.update(to_haz_curves(
+                sitecol.sids, gmfs, oq.imtls, oq.investigation_time, duration))
+    return result
+
+
+def to_haz_curves(site_ids, gmfs, imtls, investigation_time, duration):
+    """
+    :param sids: IDs of the given sites
+    :param gmfs: a list of gmf keyed by IMT
+    :param imtls: ordered dictionary {IMT: intensity measure levels}
+    :param investigation_time: investigation time
+    :param duration: investigation_time * number of Stochastic Event Sets
```

Comment:
A possible alternative could be to pass the number of stochastic event sets to the function and then calculate the duration (but the current solution might actually be better)